fv/containers/containers.go

```diff
@@ -0,0 +1,126 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containers
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/projectcalico/felix/fv/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+type Container struct {
+	Name     string
+	IP       string
+	Hostname string
+	Stop     func()
+}
+
+var containerIdx = 0
+
+var runningContainers = []*Container{}
+
+func Run(namePrefix string, args ...string) (c *Container) {
+
+	// Build unique container name and struct.
+	containerIdx++
+	c = &Container{Name: fmt.Sprintf("%v-%d-%d-", namePrefix, os.Getpid(), containerIdx)}
+
+	// Start the container.
+	log.WithField("container", c).Info("About to run container")
+	runArgs := append([]string{"run", "--name", c.Name}, args...)
+	runCmd := exec.Command("docker", runArgs...)
+	err := runCmd.Start()
+	Expect(err).NotTo(HaveOccurred())
+	c.WaitRunning(10 * time.Second)
+
+	// Remember that this container is now running.
+	runningContainers = append(runningContainers, c)
+
+	// Fill in rest of container struct.
+	c.IP = c.GetIP()
+	c.Hostname = c.GetHostname()
+	c.Stop = func() {
+		runCmd.Process.Signal(os.Interrupt)
+		c.WaitNotRunning(10 * time.Second)
+	}
+	log.WithField("container", c).Info("Container now running")
+	return
+}
+
+func (c *Container) DockerInspect(format string) string {
+	inspectCmd := exec.Command("docker", "inspect",
+		"--format="+format,
+		c.Name,
+	)
+	outputBytes, err := inspectCmd.CombinedOutput()
+	Expect(err).NotTo(HaveOccurred())
+	return string(outputBytes)
+}
+
+func (c *Container) GetIP() string {
+	output := c.DockerInspect("{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}")
+	return strings.TrimSpace(output)
+}
+
+func (c *Container) GetHostname() string {
+	output := c.DockerInspect("{{.Config.Hostname}}")
+	return strings.TrimSpace(output)
+}
+
+func (c *Container) WaitRunning(timeout time.Duration) {
+	log.Info("Wait for container to be listed in docker ps")
+	start := time.Now()
+	for {
+		cmd := exec.Command("docker", "ps")
+		out, err := cmd.CombinedOutput()
+		Expect(err).NotTo(HaveOccurred())
+		if strings.Contains(string(out), c.Name) {
+			break
+		}
+		if time.Since(start) > timeout {
+			log.Panic("Timed out waiting for container to be listed.")
+		}
+	}
+}
+
+func (c *Container) WaitNotRunning(timeout time.Duration) {
+	log.Info("Wait for container not to be listed in docker ps")
+	start := time.Now()
+	for {
+		cmd := exec.Command("docker", "ps")
+		out, err := cmd.CombinedOutput()
+		Expect(err).NotTo(HaveOccurred())
+		if !strings.Contains(string(out), c.Name) {
+			break
+		}
+		if time.Since(start) > timeout {
+			log.Panic("Timed out waiting for container not to be listed.")
+		}
+	}
+}
+
+var _ = AfterEach(func() {
+	for _, c := range runningContainers {
+		utils.Run("docker", "rm", "-f", c.Name)
```

Comment:
This is a tricky one.  For what Stop() does (i.e. interrupt the Docker process), I took my cue from your similar code; but to be absolutely sure that everything is cleaned up at the end of the test, I believe (possibly wrongly!) that `docker rm -f` is safer.

So possibly there are good reasons for these to be different.  OTOH, I wonder if there is any downside to using `docker rm -f` for Stop() as well?  If there isn't, then we could indeed make these 2 things the same.  WDYT?

Reply:
Probably best to combine them into one method that does the interrupt and then the `rm -f` as belt-and-braces.  We may need to wait on the `docker run` process to prevent it from becoming a zombie in any case.

The big advantage of using the `docker run` approach is that, if our process dies then docker spots that and kills the containers.